### PR TITLE
Multiverse sword

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Ports over TG's multiverse sword.

Basically, the multiverse sword is an item that can create alternate copies of yourself; you activate it and it binds to you--once it's bound to you, you can summon alternate copies of yourself---these copies pull from observers who have "wizard" enabled. A multiverse sword shares a cooldown with ALL other multiverse swords on the field, so keep that in mind.

The copy can end up being the same species as you or one that is completely and totally different; the equipment they may start out with may be incredibly modern or incredibly outdated. Regardless of what you roll for, you'll get a multiverse sword of your very own. Your copies will also have the same name as you.

This also ports over the alternative to "There Can Be Only one"; "There Can Be Only Me". It's basically the same thing, only you're given a multiverse sword and all access instead of getting a pre-defined set of equipment. The goal of the mode is to be the only person (and copies) of a singular type on the shuttle.

Unique Paradise Changes

I decided to tweak this item quite heavily as I really loved the concept of it; as such, I added a lot more vars to it to control aspects about it; for instance, the probability that the sword will turn out "evil" (gives the kill everyone but you and your copies objective). I also added a var to control the cooldown between uses. Additionally, I made a var to spawn in specific types of swords---so subtypes with their own unique settings will generate that same sword subtype for your clones.

Another Paradise change is that if your faction is *exactly* the same as the target you're attacking, the attack will NOT go through. This is largely there to help gut down on confusion during mass fights; you'll never be able to attack your allies, with this sword, if they were spawned in by you (likewise, you won't be able to attack them, either).

The big change however, is my favorite; if you enable the "duplicate_self" var, several additional things will activate with the sword. Instead of copies being randomly generated, they will ALWAYS be the same species as you and look identical to you---additionally, instead of just spawning with maintenance access, they'll start off with the same, exact, access as you have.

Lastly, any items currently equipped to your character (minus hands) will also be duplicated over to the copy--this goes for ALL equipment (Please note, that items inside storage based items like backpacks and belts will NOT be duplicates).

There's two pre-edited types, for now:

- Pure evil multiverse sword (for Only Me mode).
 - Always grants the hijack objective
- A Pike
 - Has duplicate_self enabled by default. Additionally has a lower clone spawn CD and will always grant hijack.

Pics:
noyoudidnt

And yes, before anyone asks, the self-duplication was 100% inspired by this [guy](http://cdn.dota2.com/apps/dota2/images/heroes/phantom_lancer_vert.jpg), though I suspect that was already obvious.

Related
 - Makes Thunderdome.ogg play whenever Only One or Only Me is started.

### CODE REVIEW

**Comment by KasparoVy** [comment]
Holy god-tier port

Looks like it's going to be wicked fun, awesome work bringing this over! The additional features will be hella handy

**Comment by theColdflame** [comment]
So does this item appear in normal play or is it an event?

Regardless, this seems really neat.

**Comment by Fox-McCloud** [comment]
9/10 times it's only ever going to be seen with "Only Me". It's not accessible, beyond that.

Why did I spent so much time customizing it, then? I dunno. I like PL =p

**Comment by Aurorablade** [comment]
it WAS a wizard summon if i recall..but was so OP it had to be taken out of being a wizard summon...But we have silly silly admins who like to drive an Adminbus here.

**Comment by Fox-McCloud** [comment]
>it WAS a wizard summon if i recall..but was so OP it had to be taken out of being a wizard summon

This, pretty much; I mean, if you just huddle on some corner of the station with it where there's no camera vision and spam the everloving crap out of it for 7 minutes, you can pretty much just run the station over with no problem, not to mention that when you kill someone you can pull them onto your side after you kill them.

File: code/game/gamemodes/objective.dm [hunk 1]
```diff
@@ -347,6 +347,36 @@ datum/objective/hijack
...
+
+		var/area/A = locate(/area/shuttle/escape/centcom)
+
+		for(var/mob/living/player in player_list) //Make sure nobody else is onboard
+			if(player.mind && player.mind != owner)
+				if(player.stat != DEAD)
+					if(istype(player, /mob/living/silicon))
+						continue
+					if(get_area(player) == A)
+						if(player.real_name != owner.current.real_name && !istype(get_turf(player.mind.current), /turf/simulated/shuttle/floor4))
```
Comment (tigercat2000) [inline code/game/gamemodes/objective.dm]: this.. is a rather flimsy check, dont'cha think?

File: code/game/gamemodes/wizard/artefact.dm [hunk 2]
```diff
@@ -213,3 +213,383 @@
...
+	..()
+	multiverse |= src
+
+
+/obj/item/weapon/multisword/Destroy()
+	multiverse.Remove(src)
+	return ..()
+
+/obj/item/weapon/multisword/attack(mob/living/M as mob, mob/living/user as mob)  //to prevent accidental friendly fire or out and out grief.
+	if(M.faction == user.faction)
```
Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: I'm not entirely sure if lists behave properly with being compared by `==`

**Comment by Spacemanspark** [comment]
:+1: Jane Doe's and Kei's running everywhere. :^)

File: code/game/gamemodes/wizard/artefact.dm [hunk 3]
```diff
@@ -213,3 +213,383 @@
...
+	M.key = C.key
+	M.mind.name = usr.real_name
+	M << "<B>You are an alternate version of [usr.real_name] from another universe! Help them accomplish their goals at all costs.</B>"
+	M.faction = list("[usr.real_name]")
+	if(duplicate_self)
+		M.set_species(usr.get_species()) //duplicate the sword user's species.
+	else
+		if(prob(50))
+			var/list/all_species = list("Human","Unathi","Skrell","Tajaran","Kidan","Golem","Diona","Machine","Slime People","Grey","Vulpkanin")
+			M.set_species(pick(all_species))
```
Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: This can behave oddly after using `C.prefs.copy_to`

File: code/game/gamemodes/wizard/artefact.dm [hunk 4]
```diff
@@ -213,3 +213,383 @@
...
+	M.real_name = usr.real_name //this is clear down here in case the user happens to become a golem; that way they have the proper name.
+	M.name = usr.real_name
+	if(duplicate_self)
+		M.dna = usr.dna.Clone()
+		M.UpdateAppearance()
+		domutcheck(M, null)
+	M.update_body()
```
Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: I believe you need to force-regenerate limbs after setting DNA, or the limbs don't take on the DNA.

File: code/game/gamemodes/wizard/artefact.dm [hunk 5]
```diff
@@ -213,3 +213,383 @@
...
+				M.equip_to_slot_or_del(new /obj/item/clothing/mask/surgical(M), slot_wear_mask)
+				M.equip_to_slot_or_del(new /obj/item/clothing/head/welding(M), slot_head)
+				M.equip_to_slot_or_del(new /obj/item/device/radio/headset(M), slot_l_ear)
+				M.equip_to_slot_or_del(new /obj/item/clothing/suit/apron(M), slot_wear_suit)
+				M.equip_to_slot_or_del(new /obj/item/weapon/kitchen/utensil/knife(M), slot_l_store)
+				M.equip_to_slot_or_del(new /obj/item/weapon/scalpel(M), slot_r_store)
+				M.equip_to_slot_or_del(sword, slot_r_hand)
+				for(var/obj/item/carried_item in M.contents)
+					if(!istype(carried_item, /obj/item/weapon/implant))
+						carried_item.add_blood(M)
```
Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: Obviously this is a real proc, or it wouldn't have compiled; But this is -tg-'s standard for adding blood to an object, I question if it actually works for what you want to do, since we have such a massively different system.

File: code/game/gamemodes/wizard/artefact.dm [hunk 6]
```diff
@@ -213,3 +213,383 @@
...
+				M.equip_to_slot_or_del(sword, slot_r_hand)
+
+			if("soviet")
+				M.equip_to_slot_or_del(new /obj/item/clothing/head/hgpiratecap(M), slot_head)
+				M.equip_to_slot_or_del(new /obj/item/clothing/shoes/combat(M), slot_shoes)
+				M.equip_to_slot_or_del(new /obj/item/clothing/gloves/combat(M), slot_gloves)
+				M.equip_to_slot_or_del(new /obj/item/device/radio/headset(M), slot_l_ear)
+				M.equip_to_slot_or_del(new /obj/item/clothing/suit/hgpirate(M), slot_wear_suit)
+				M.equip_to_slot_or_del(new /obj/item/clothing/under/soviet(M), slot_w_uniform)
+				M.equip_to_slot_or_del(sword, slot_r_hand)
```
Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: Also, all of these should really be `M.equip_or_collect(`, or it's possible for any of these to be abruptly deleted if they don't fit in the right slot.

File: code/game/gamemodes/wizard/artefact.dm [hunk 7]
```diff
@@ -213,3 +213,383 @@
...
+			W.icon_state = duplicated_id.icon_state
+		else
+			W.access += access_maint_tunnels
+			W.icon_state = "centcom"
+	else
+		W.access += access_maint_tunnels
+		W.icon_state = "centcom"
+	W.assignment = "Multiverse Traveller"
+	W.registered_name = M.real_name
+	W.update_label(M.real_name)
```
Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: Don't forget to run `W.SetOwnerInfo()`, or the ID examine menu won't generate.

File: code/game/gamemodes/wizard/artefact.dm (hunk 2 above)
Comment (RemieRichards) [inline code/game/gamemodes/wizard/artefact.dm]: Given lists are objects (Set two list vars to the same list reference and they will have the same list and update whenever the other var's ref to it updates) this check only ensures that the list object is the same, and since it's created on New() for each mob, it won't be (Unless fox's code here copies the same list object over, in which case -maybe-)

tl;dr bad check,

File: code/game/gamemodes/wizard/artefact.dm (hunk 6 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: In this case, they need to be del, or it's going to wind up doing weird things if you happen to roll for golem, for instance.

File: code/game/gamemodes/wizard/artefact.dm (hunk 4 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: Which could lead to?

File: code/game/gamemodes/wizard/artefact.dm (hunk 3 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: Such as?

File: code/game/gamemodes/wizard/artefact.dm (hunk 6 above)
Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: equip_to_slot_or_del will delete the input item if it doesn't fit into the slot. If they roll for golem, they will simply not get any of their clothing.

File: code/game/gamemodes/objective.dm (hunk 1 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/objective.dm]: Not really sure what you mean it's checking that real name is identical--considering how difficult this is to change (and AFAIK, the only way that players can alter this is via genetics), I'm not really sure if that's a a problem. The floor check is to make an exception for the brig area on the shuttle.

File: code/game/gamemodes/wizard/artefact.dm (hunk 4 above)
Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: Limbs generate their icon based on their DNA.

File: code/game/gamemodes/wizard/artefact.dm (hunk 3 above)
Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: Well, it might have been fixed, can't remember- but set_species didn't properly null the hair style and facial hair style last time I checked

File: code/game/gamemodes/wizard/artefact.dm (hunk 4 above)
Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: Wait, no, disregard that last comment, limbs do not generate their icons based on their DNA- but they do generate their skin tone/color off of it.

File: code/game/gamemodes/wizard/artefact.dm (hunk 5 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: Both systems utilize the same proc names, and both systems effectively do the same thing--the difference is, Bay's has some extra stuff for species' blood color.

File: code/game/gamemodes/wizard/artefact.dm (hunk 6 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: Yes, I'm perfectly aware of how equip or del and equip or collect works, and in this case, the items should be deleted instead of collected.

File: code/game/gamemodes/wizard/artefact.dm (hunk 3 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: I didn't notice any Tajarans or Unathi or such with human hair styles with this, or anything like that--this was tested primarily with human characters; when a human was generated, it had the user settings' hair--when it wasn't, they were bald.

File: code/game/gamemodes/wizard/artefact.dm (hunk 4 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: duplicate copies are the only ones that inherit DNA directly---with those? They were definitely receiving the proper limb color.

Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: Oh, UpdateApperance already calls force_regenerate_limbs. Nevermind

File: code/game/gamemodes/wizard/artefact.dm (hunk 7 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: Forgot about that.

File: code/game/gamemodes/wizard/artefact.dm (hunk 2 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: Yeah, was a last minute add--will do something else instead.

File: code/game/gamemodes/wizard/artefact.dm (hunk 5 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: Update: This does, in fact, work properly.

Comment (tigercat2000) [inline code/game/gamemodes/wizard/artefact.dm]: alright great

**Comment by tigercat2000** [comment]
I am now going to find out if deleting OP posts on line diffs will delete the entire chain! because I pointed out multiple things while tired and it's borderline spammy.

**Comment by tigercat2000** [comment]
Update: Totes does.

**Comment by Fox-McCloud** [comment]
Last  few tweaks.

- Did an alternate safety check
 - Tested and works.
- Ensured that Tajarans and Unathi properly spawn in with sandals.
- Ensured that icons are updated after equipping when only me is activated.

File: code/game/gamemodes/wizard/artefact.dm (hunk 7 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: addressed.

File: code/game/gamemodes/wizard/artefact.dm (hunk 5 above)
Comment (KasparoVy) [inline code/game/gamemodes/wizard/artefact.dm]: So if you create a duplicate of yourself as a Vox with the sword and they come out Vox as well, then you cut them, will they bleed teal as per usual? Or will it be human red?

Comment (Fox-McCloud) [inline code/game/gamemodes/wizard/artefact.dm]: They would have blue blood still.

That said, it's impossible to have a Vox clone unless the sword has duplicate_self enabled.

**Comment by Fox-McCloud** [comment]
@Markolie @TheDZD 

This is ready to be merged (by the time you guys see this), unless you notice any other problems with it.

**Comment by Regen1** [comment]
"We have strength in numbers, what did YOU have?"
http://hydra-media.cursecdn.com/dota2.gamepedia.com/e/ee/Plance_kill_09.mp3
